url47fcfc0133_baliosian_tau_fst_tgz-pJ8-uy_edu_fing_mina_fsa_test_old_TestComposition1J8/src/uy/edu/fing/mina/fsa/tffst/P.java

```diff
@@ -0,0 +1,410 @@
+				workingTFinSE = wSEiter.next();
+			  if (currentTFinSE != null && workingTFinSE != null)
+				newSE.add(workingTFinSE.andSimple(currentTFinSE));
+			  else if (currentTFinSE == null)
+				newSE.add(workingTFinSE);
+			  else if (workingTFinSE == null)
+				newSE.add(currentTFinSE);
+			}
+			outWP.arrivingTFs = newSE.clone();
+		  }
+		}
+
+		if (!longestPosfixState.isEmpty()) {
+		  for (ElementOfP eoup : prep)
+			for (ElementOfP eovp : longestPosfixState)
+			  if (eoup.state.equals(eovp.state)) {
+				TfString atf = outWP.arrivingTFs.clone();
+				atf.addAll(eovp.arrivingTFs);
+				outp.add(new ElementOfP(eoup.state, atf));
+			  }
+		} else {
+		  for (ElementOfP eoup : prep)
+			outp.add(new ElementOfP(eoup.state, (TfString) outWP.arrivingTFs.clone()));
+		}
+		this.p = outp.p;
+	  }
+	} catch (CloneNotSupportedException e) {
+	  e.printStackTrace();
+	}
+  }
+
+  /**
+   * retrieves the P state with the longest common arrivingTf tail. 
+   * it assumes that this.p and p in the visited states has been simplified beforehand by simplifyTargetByState(). 
+   * this means that there is no two target pair with same state.    
+   * 
+   * @param visitedNewStates
+   * @return
+   */
+  
+  
+  P[] longestPosfixState(Map<P, State> visitedNewStates) {
+    P[] candidate = new P[2];
+    candidate[0] = new P();
+    candidate[1] = new P();
+    int candidateMaxLen = -1;
+
+    for (P visitedP : visitedNewStates.keySet()) {
+      Map<State, TfString[] > lastCandidate = new HashMap<State, TfString[]>();
+      boolean vpmatch = true;
+      if (visitedP.size() == p.size()) {
+        int eovpMaxLen = 0;
+		for (ElementOfP eop : p) {
+		  boolean eovpmatch = false;
+		  for (ElementOfP eovp : visitedP) {
+			if (eop.state.equals(eovp.state)) {
+//			  if (eovp.arrivingTFs.size() <= eop.arrivingTFs.size()) {
+				if (eovp.arrivingTFs.isEpsilon()) {
+				  eovpmatch = true;
+				  eovpMaxLen = Math.max(eovpMaxLen, 0);
+				  lastCandidate.put(eop.state, new TfString[] {new TfString(), new TfString()});
+				} else if (!eop.arrivingTFs.isEpsilon()) {
+				  for (int i = 1 ; 
+					  (i <= eovp.arrivingTFs.size() && i <= eop.arrivingTFs.size() && 
+					  eop.arrivingTFs.subList(eop.arrivingTFs.size()-i, eop.arrivingTFs.size())
+					  .equals(eovp.arrivingTFs.subList(eovp.arrivingTFs.size()-i, eovp.arrivingTFs.size())));
+					  i++) {
+					lastCandidate.put(eop.state, new TfString[] {new TfString(eovp.arrivingTFs.subList(0, eovp.arrivingTFs.size()-i)), new TfString(eop.arrivingTFs.subList(eop.arrivingTFs.size()-i, eop.arrivingTFs.size()))});
+					eovpmatch = true;
+					eovpMaxLen = Math.max(eovpMaxLen, i);
+				  }
+				}
+			//  }
+			}
+		  }
+		  vpmatch = vpmatch && eovpmatch;
+		}
+        if (vpmatch && candidateMaxLen < eovpMaxLen) {
+          for (State s : lastCandidate.keySet()) {
+			candidate[0].add(new ElementOfP(s, lastCandidate.get(s)[0]));
+			candidate[1].add(new ElementOfP(s, lastCandidate.get(s)[1]));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.